keep-runtime/src/workload.rs

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: Apache-2.0
+
+/// The error codes of workload execution.
+#[derive(Debug)]
+pub enum Error {
+    /// import module not found
+    ImportModuleNotFound(String),
+    /// import field not found
+    ImportFieldNotFound(String, String),
+    /// export not found
+    ExportNotFound,
+    /// call failed
+    CallFailed,
+    /// runtime error
+    RuntimeError,
+    /// I/O error
+    IoError(std::io::Error),
+}
+
+impl From<std::io::Error> for Error {
+    fn from(err: std::io::Error) -> Self {
+        Self::IoError(err)
+    }
+}
+
+/// Result type used throughout the library.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Runs a WebAssembly workload.
+pub fn run(
+    reader: &mut impl std::io::Read,
+    args: Box<[String]>,
+    envs: Box<[(String, String)]>,
+) -> Result<Box<[wasmtime::Val]>> {
```

Comment:
```suggestion
pub fn run<A: AsRef<str>, E: AsRef<str>>(
    reader: impl AsRef<[u8]>,
    args: impl IntoIterator<Item = A>,
    envs: impl IntoIterator<Item = E>,
) -> Result<Box<[wasmtime::Val]>> {
```

Reply:
It was a bit complicated because `envs` iterates over tuples. I changed the signature to:
```rust
pub fn run<S, Args, Envs>(
    bytes: impl AsRef<[u8]>,
    args: Args,
    envs: Envs,
) -> Result<Box<[wasmtime::Val]>>
where
    Args: IntoIterator<Item = S>,
    Envs: IntoIterator<Item = (S, S)>,
    S: AsRef<[u8]>;
```